### PR TITLE
Extract examples from documentation and run tests on it

### LINKED ISSUE #2552: Parse the sources from documentation for the fuzzer tests
Extend `scripts/isolate_tests.py` to do that.

Search for lines indented by 4 spaces and consider the lines following `pragma solidity` as source code until the indentation is less than 4 spaces.

If the comment `// This will not compile` is placed above the version pragma, ignore that source.

### PR DESCRIPTION
**Pull request by axic**
Fixes #2552.

### CODE REVIEW

**Comment by axic** [comment]
Only found three failures:

```
test_32515c3cbd76ec0613826867331326c0b3196b54c23fc45f812c2f2b596d5158.sol:12:35: Error: Expected primary expression.
            return MappingExample(<address>).balances(this);
                                  ^
test_576ffe0cff04b5e42f0c57f505ffb33d48e8aaf83a391229dc1d1b6f89c5ced0.sol:3:44: Error: Expected primary expression.
        function f(uint key, uint value) { ... }
                                           ^
test_aff284277f37146ad6c8cff20ef833738536497e65b029650ff15a25f08cb5f8.sol:9:1: Error: Function, variable, struct or modifier declaration expected.

^

```

File: test/cmdlineTests.sh [hunk 1]
```diff
@@ -49,6 +49,30 @@ do
...
+    set -e
+    cd "$REPO_ROOT"
+    REPO_ROOT=$(pwd) # make it absolute
+    cd "$TMPDIR"
+    "$REPO_ROOT"/scripts/isolate_tests.py "$REPO_ROOT"/docs/ docs
+    for f in *.sol
+    do
+        echo "trying $f"
+        set +e
+        output=$("$SOLC" "$f" 2>&1)
```
Comment (axic) [inline test/cmdlineTests.sh]: This is a copy of the previous test block. @chriseth if you could transform it into a bash function that could simplify it.

File: scripts/isolate_tests.py [hunk 2]
```diff
@@ -34,15 +34,60 @@ def extract_cases(path):
...
+    return tests
 
 def write_cases(tests):
     for test in tests:
         open('test_%s.sol' % hashlib.sha256(test).hexdigest(), 'wb').write(test)
 
 if __name__ == '__main__':
     path = sys.argv[1]
+    docs = False
+    if len(sys.argv) > 2 and sys.argv[2] == 'docs':
```
Comment (chriseth) [inline scripts/isolate_tests.py:83]: Why not just do both all the time?

File: test/cmdlineTests.sh (hunk 1 above)
Comment (chriseth) [reply test/cmdlineTests.sh]: Do we want these to be more strict, e.g. require no warnings?

File: scripts/isolate_tests.py (hunk 2 above)
Comment (axic) [reply scripts/isolate_tests.py:83]: They require a different source directory (argv[1]).

**Comment by chriseth** [comment]
Trying to convert to a function.

File: test/cmdlineTests.sh (hunk 1 above)
Comment (axic) [reply test/cmdlineTests.sh]: I guess we could aim for that (with the exception of the pre-release compiler warning).

**Comment by axic** [comment]
Fails on paths:

```
Compiling all examples from the documentation...
trying test_12a2345f60bc89656aca0f44ee3327794490078cff6f15cdcb652a653c8da530.sol
scripts/../test/cmdlineTests.sh: line 64: scripts/../test/../build/solc/solc: No such file or directory
```

**Comment by chriseth** [comment]
Actually I think we should first merge https://github.com/ethereum/solidity/pull/2522

File: test/cmdlineTests.sh (hunk 1 above)
Comment (chriseth) [reply test/cmdlineTests.sh]: Hm, but I won't add that to this pull request for now...

File: test/cmdlineTests.sh [hunk 3]
```diff
@@ -28,27 +28,86 @@
 
 set -e
 
-REPO_ROOT="$(dirname "$0")"/..
+REPO_ROOT=$(cd "$REPO_ROOT" && pwd)
```
Comment (axic) [inline test/cmdlineTests.sh]: How is this working?

File: docs/contracts.rst [hunk 4]
```diff
@@ -914,7 +927,9 @@ Contract functions can lack an implementation as in the following example (note
         function utterance() returns (bytes32);
     }
 
-Such contracts cannot be compiled (even if they contain implemented functions alongside non-implemented functions), but they can be used as base contracts::
+    // Such contracts cannot be compiled (even if they contain
+    // implemented functions alongside non-implemented functions),
+    // but they can be used as base contracts::
```
Comment (axic) [inline docs/contracts.rst]: I think we're better off keeping this text unchanged and duplicating the `Feline` contract here.

File: docs/security-considerations.rst [hunk 5]
```diff
@@ -179,11 +179,7 @@ Never use tx.origin for authorization. Let's say you have a wallet contract like
         }
     }
 
-Now someone tricks you into sending ether to the address of this attack wallet:
-
-::
-
-    pragma solidity ^0.4.0;
+    // Now someone tricks you into sending ether to the address of this attack wallet:
```
Comment (axic) [inline docs/security-considerations.rst]: I think we could keep the original text and add the interface of `transferTo`.

File: test/cmdlineTests.sh (hunk 3 above)
Comment (chriseth) [reply test/cmdlineTests.sh]: `$(...)` executes the command(s) inside the parantheses and returns the output of that command as a value. `&&` just means "only execute the next command if the previous command succeeded" i.e. exactly as the `&&` in C. `pwd` prints the (absolute) path to the current directory.

Comment (axic) [reply test/cmdlineTests.sh]: I mean it references `REPO_ROOT` which is just an empty string at that point, so effectively it does `cd` and doesn't changes the current directory.

Comment (chriseth) [reply test/cmdlineTests.sh]: Ah, now I see what you mean! Sorry, will fix it.

**Comment by chriseth** [comment]
Addressed all comments.

File: docs/security-considerations.rst [hunk 6]
```diff
@@ -179,7 +179,11 @@ Never use tx.origin for authorization. Let's say you have a wallet contract like
         }
     }
 
-    // Now someone tricks you into sending ether to the address of this attack wallet:
+Now someone tricks you into sending ether to the address of this attack wallet::
+
+    interface TxUserWallet {
```
Comment (axic) [inline docs/security-considerations.rst]: pragma is missing here

**Comment by axic** [comment]
Changing this to fix the above.

**Comment by axic** [comment]
@chriseth this is ready from my side

**Comment by axic** [comment]
@chriseth you need to approve this in order to merge

**Pull request merged**